### ISSUE TITLE
[Bug]: signal SIGSEGV in AOE

### ISSUE THREAD

**Issue by lni** [matrixorigin/matrixone#1350]
### Is there an existing issue for the same bug?

- [X] I have checked the existing issues.

### Environment

```markdown
- Version or commit-id (e.g. v0.1.0 or 8b23a93): 77f928d5f686b88b5bde0144ea49d5d3ab99ddfb from https://github.com/jiangxinmeng1/matrixone-1
- Hardware parameters:
- OS type:
- Others:
```

### Actual Behavior

```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x2 addr=0x38 pc=0x105de8ac0]

goroutine 518 [running]:
github.com/matrixorigin/matrixone/pkg/vm/engine/aoe/storage/layout/dataio.(*BlockWriter).executeIVecs(0x14010f6e8c0)
  /Users/lni/src/matrixone/pkg/vm/engine/aoe/storage/layout/dataio/blkwriter.go:228 +0x310
github.com/matrixorigin/matrixone/pkg/vm/engine/aoe/storage/layout/dataio.(*BlockWriter).Execute(0x14010f6e8c0)
  /Users/lni/src/matrixone/pkg/vm/engine/aoe/storage/layout/dataio/blkwriter.go:190 +0x38
github.com/matrixorigin/matrixone/pkg/vm/engine/aoe/storage/layout/dataio.(*TransientBlockFile).Sync(0x14010f2fa00, {0x106532600, 0x140113638c0}, 0x140124548a0)
  /Users/lni/src/matrixone/pkg/vm/engine/aoe/storage/layout/dataio/tblk.go:279 +0x3b4
github.com/matrixorigin/matrixone/pkg/vm/engine/aoe/storage/mutation.blockFlusher.flush(...)
  /Users/lni/src/matrixone/pkg/vm/engine/aoe/storage/mutation/blk.go:32
github.com/matrixorigin/matrixone/pkg/vm/engine/aoe/storage/mutation.(*MutableBlockNode).Flush(0x1400ae73770)
  /Users/lni/src/matrixone/pkg/vm/engine/aoe/storage/mutation/blk.go:102 +0x3c0
github.com/matrixorigin/matrixone/pkg/vm/engine/aoe/storage/layout/table/v1.(*tableAppender).Flush(0x1401232c0f0)
  /Users/lni/src/matrixone/pkg/vm/engine/aoe/storage/layout/table/v1/appender.go:66 +0x1cc
github.com/matrixorigin/matrixone/pkg/vm/engine/aoe/storage/db.(*DB).DoFlushTable(0x14000148690, 0x14011363140)
  /Users/lni/src/matrixone/pkg/vm/engine/aoe/storage/db/impl.go:58 +0x90
github.com/matrixorigin/matrixone/pkg/vm/engine/aoe/storage/db.(*DB).DoFlushDatabase(0x14000148690, 0x14011b1f280)
  /Users/lni/src/matrixone/pkg/vm/engine/aoe/storage/db/impl.go:43 +0xc4
github.com/matrixorigin/matrixone/pkg/vm/engine/aoe/storage/db.(*DB).FlushDatabase(0x14000148690, {0x105fbcef8, 0x2})
  /Users/lni/src/matrixone/pkg/vm/engine/aoe/storage/db/db.go:123 +0xb4
github.com/matrixorigin/matrixone/pkg/vm/driver/aoe.(*Storage).Sync(0x1400038fa70, {0x14011de1560, 0x1, 0x1})
  /Users/lni/src/matrixone/pkg/vm/driver/aoe/storage.go:64 +0x78
github.com/matrixorigin/matrixcube/raftstore.(*replicaCreator).maybeSaveMetadata.func1({0x10653f380, 0x1400038fa70}, {0x14011df1400, 0x1, 0x1})
  /Users/lni/src/matrixcube/raftstore/replica_create.go:161 +0x1e8
github.com/matrixorigin/matrixcube/raftstore.doWithShardsByGroupID(0x1400a718d88, 0x1400a718d78, {0x14011df1300, 0x1, 0x1})
  /Users/lni/src/matrixcube/raftstore/replica_create.go:254 +0xd4
github.com/matrixorigin/matrixcube/raftstore.(*replicaCreator).maybeSaveMetadata(0x1400a718f50, {0x14011df1300, 0x1, 0x1})
  /Users/lni/src/matrixcube/raftstore/replica_create.go:130 +0x80
github.com/matrixorigin/matrixcube/raftstore.(*replicaCreator).create(0x1400a718f50, {0x1400a719050, 0x1, 0x1})
  /Users/lni/src/matrixcube/raftstore/replica_create.go:93 +0x94
github.com/matrixorigin/matrixcube/raftstore.(*store).doDynamicallyCreate(0x14000602820, {0x38, {0x14011de1410, 0xc, 0x10}, {0x14011de1420, 0xc, 0x10}, {0x3, 0x0, ...}, ...})
  /Users/lni/src/matrixcube/raftstore/store_route_handler.go:36 +0x168
github.com/matrixorigin/matrixcube/raftstore.(*store).startRouter.func1({0x38, {0x14011de1410, 0xc, 0x10}, {0x14011de1420, 0xc, 0x10}, {0x3, 0x0, {}, ...}, ...})
  /Users/lni/src/matrixcube/raftstore/store.go:271 +0x4c
github.com/matrixorigin/matrixcube/raftstore.(*defaultRouter).updateShardLocked(0x14000362700, {0x1400eff9cc0, 0x47, 0x50}, 0x0, 0x0, 0x1)
  /Users/lni/src/matrixcube/raftstore/router.go:313 +0x1d0
github.com/matrixorigin/matrixcube/raftstore.(*defaultRouter).handleEvent(0x14000362700, {0x23, 0x4, 0x0, 0x1400f26edc0, 0x0, 0x0, 0x0, {}, {0x0, ...}, ...})
  /Users/lni/src/matrixcube/raftstore/router.go:276 +0x4b0
github.com/matrixorigin/matrixcube/raftstore.(*defaultRouter).eventLoop(0x14000362700)
  /Users/lni/src/matrixcube/raftstore/router.go:249 +0x98
github.com/lni/goutils/syncutil.(*Stopper).runWorker.func1(0x14000088ba0, 0x1400035b4a0, {0x0, 0x0}, 0x14009f9cba0)
  /Users/lni/go/pkg/mod/github.com/lni/goutils@v1.3.0/syncutil/stopper.go:79 +0xc0
created by github.com/lni/goutils/syncutil.(*Stopper).runWorker
  /Users/lni/go/pkg/mod/github.com/lni/goutils@v1.3.0/syncutil/stopper.go:74 +0x88
FAIL  github.com/matrixorigin/matrixone/pkg/vm/driver/test  21.121s
FAIL
```

### Expected Behavior

No error.

### Steps to Reproduce

```markdown
Run TestSnapshot from 77f928d5f686b88b5bde0144ea49d5d3ab99ddfb from https://github.com/jiangxinmeng1/matrixone-1
```

### Additional information

data folder

[test2.tar.gz](https://github.com/matrixorigin/matrixone/files/7693609/test2.tar.gz)

**Comment by XuPeng-SH** [comment]
> Root Cause

Right now, the shard group meta info is saved into underlying AOE table when the shard is created. Since AOE shares WAL with external raft log, all table data is flushed asynchronously. If a crash (or shutdown) happens before the data is flushed, it can be replayed from WAL. 

In the current process, when restart a Shard in CUBE (log consensus layer), before replaying the log, it need to load the shard meta info from underlying AOE engine and then replay the log. However the AOE engine is sharing WAL with consensus layer and it may need to replay the log first to restore it to the correct state. This is a situation of interdependence.

> Solution

One of the simplest solutions is that when creating shards, operations on AOE are synchronized and deterministic.

> Why is there no such problem on the pebble engine

Pebble has an WAL and it replays the log from WAL to restore it to the latest state. It doesn't rely on the log from the consensus layer.

**Comment by XuPeng-SH** [comment]
should be fixed by #1389
